### PR TITLE
bug/analytics reset no longer clears anonoymousId

### PR DESCRIPTION
**Pull request by silesky**
There was a race condition where the segment "normalize" plugin would asyncronously instantiate / overwrite the anonymousId field in browser storage right before the http request, thus making it impossible to call analytics.reset() after a typical method call. (for example, in a .track call). Moved this to the EventFactory so at the very least this browser storage action happens syncronously.

The only behavioral change is that if someone mutates event.anonymousId in a plug-in, it's not going to also implicitly update the global user anonymousId when that event gets sent.

this seems... fine to me? I wish we could go a step further and force everyone to call "setAnonymousId" explicitly if they wanted to override the global anonymousId for all events.

### CODE REVIEW

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 1]
```diff
@@ -322,6 +322,45 @@ describe('Event Factory', () => {
...
+
+      test('if two different anonymousIds are passed, should use one on the event', () => {
+        const track = factory.track('Order Completed', shoes, {
+          anonymousId: 'bar',
+          context: {
+            anonymousId: 'foo',
+          },
+        })
+        expect(track.context?.anonymousId).toBe('bar')
+        expect(track.anonymousId).toBe('bar')
```
Comment (silesky) [inline packages/browser/src/core/events/__tests__/index.test.ts:351]: Failure!

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 2]
```diff
@@ -322,6 +322,45 @@ describe('Event Factory', () => {
...
+      })
+
+      test('should set an anonymousId passed from the context on the event', () => {
```
Comment (silesky) [inline packages/browser/src/core/events/__tests__/index.test.ts:354]: Failure

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 3]
```diff
@@ -322,6 +322,45 @@ describe('Event Factory', () => {
...
+        const track = factory.track('Order Completed', shoes, {
+          context: {
+            anonymousId: 'foo',
+          },
+        })
+        expect(track.context?.anonymousId).toBe('foo')
+        expect(track.anonymousId).toBe('foo')
```
Comment (silesky) [inline packages/browser/src/core/events/__tests__/index.test.ts:361]: Failure

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 4]
```diff
@@ -322,6 +322,45 @@ describe('Event Factory', () => {
...
       })
     })
+
+    describe('anonymousId', () => {
+      test('accepts an anonymousId', () => {
+        const track = factory.track('Order Completed', shoes, {
+          anonymousId: 'foo',
+        })
+        expect(track.anonymousId).toBe('foo')
+        expect(track.context?.anonymousId).toBe('foo')
```
Comment (silesky) [inline packages/browser/src/core/events/__tests__/index.test.ts:332]: Failure

File: packages/browser/src/core/analytics/__tests__/integration.test.ts [hunk 5]
```diff
@@ -224,6 +224,19 @@ describe('Analytics', () => {
...
+      const analytics = new Analytics({ writeKey: '' })
+      const getAnonId = () =>
+        retrieveStoredData({
+          cookieNames: ['ajs_anonymous_id'],
+          localStorageKeys: ['ajs_anonymous_id'],
+        }).ajs_anonymous_id
+      void analytics.track('foo')
+      expect(getAnonId()).toBeTruthy()
+      analytics.reset()
+      expect(getAnonId()).toBeFalsy()
```
Comment (chrisradek) [inline packages/browser/src/core/analytics/__tests__/integration.test.ts:237]: Since the race condition was that the `track` call would eventually set anonId again, I think checking that it is falsey after the track call completes would cover it (and fail without these changes)

Comment (silesky) [reply packages/browser/src/core/analytics/__tests__/integration.test.ts]: I actually moved the test to a different file, tested both failure and success condition!

File: packages/browser/src/test-helpers/browser-storage.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import cookie from 'js-cookie'
```
Comment (silesky) [inline packages/browser/src/test-helpers/browser-storage.ts:1]: Refactored to make type safe

File: packages/browser/src/core/events/__tests__/index.test.ts [hunk 7]
```diff
@@ -322,6 +322,48 @@ describe('Event Factory', () => {
         innerProp: '👻',
       })
     })
+
+    describe.skip('anonymousId', () => {
```
Comment (chrisradek) [inline packages/browser/src/core/events/__tests__/index.test.ts:326]: These tests currently don't work? Are they regressions or existing issues?

Comment (silesky) [reply packages/browser/src/core/events/__tests__/index.test.ts:326]: Not regressions, they are just existing edge cases / undefined behavior that I uncovered when working on this bug.

**Review by chrisradek — APPROVED** [review]
:fire:

File: packages/browser/src/core/events/index.ts [hunk 8]
```diff
@@ -205,6 +205,15 @@ export class EventFactory {
   }
 
   public normalize(event: SegmentEvent): SegmentEvent {
+    const anonymousIdOverride = event.options?.anonymousId
```
Comment (pooyaj) [inline packages/browser/src/core/events/index.ts:208]: maybe this is more readble?  
```
event.options?.anonymousId && this.user.anonymousId(event.options.anonymousId)
```
as `event.anonymousId = id` seems to be a noop and happens regardless of this line of code.

**Review by pooyaj — APPROVED** [review]
looks great!

File: packages/browser/src/core/events/index.ts (hunk 8 above)
Comment (silesky) [reply packages/browser/src/core/events/index.ts:208]: Good catch!
